domain/application/service/application.go

```diff
@@ -1257,23 +1268,80 @@ func (s *Service) GetApplicationsWithPendingCharmsFromUUIDs(ctx context.Context,
 
 // ReserveCharmDownload reserves a charm download slot for the specified
 // application. If the charm is already being downloaded, the method will
-// return [applicationerrors.AlreadyDownloadingCharm]. The charm download
+// return [applicationerrors.CharmAlreadyAvailable]. The charm download
 // information is returned which includes the charm name, origin and the
 // digest.
 func (s *Service) ReserveCharmDownload(ctx context.Context, appID coreapplication.ID) (application.CharmDownloadInfo, error) {
 	if err := appID.Validate(); err != nil {
 		return application.CharmDownloadInfo{}, internalerrors.Errorf("application ID: %w", err)
 	}
 
-	return application.CharmDownloadInfo{}, errors.NotImplementedf("ReserveCharmDownload")
+	return s.st.ReserveCharmDownload(ctx, appID)
 }
 
 // ResolveCharmDownload resolves the charm download slot for the specified
 // application. The method will update the charm with the specified charm
 // information.
+// This returns [applicationerrors.CharmNotResolved] if the charm UUID isn't
+// the same as the one that was reserved.
 func (s *Service) ResolveCharmDownload(ctx context.Context, appID coreapplication.ID, resolve application.ResolveCharmDownload) error {
 	if err := appID.Validate(); err != nil {
 		return internalerrors.Errorf("application ID: %w", err)
 	}
-	return errors.NotImplementedf("ResolveCharmDownload")
+
+	// Although, we're resolving the charm download, we're calling the
+	// reserve method to ensure that the charm download slot is still valid.
+	// This has the added benefit of returning the charm hash, so that we can
+	// verify the charm download. We don't want it to be passed in the resolve
+	// charm download, in case the caller has the wrong hash.
+	info, err := s.ReserveCharmDownload(ctx, appID)
```

Comment:
Because we don't modify state, we're good for retries :dagger: This is by design. I'll rename the prior method to indicate it's just a read. There is a lot of code that expects retries. Hence, the different error checking for already available and already resolved.

So to answer your question, because there is only 1 write transaction, and the objectstore is idempotent for hash conflicts (assuming you handle the error correctly) then multiple retries can be applied.